benchmark/Cpp/Savina/src/parallelism/RadixSort.lf

```diff
@@ -0,0 +1,197 @@
+        } else {
+            orderingArray[pos] = (inValue.get());
+            pos += 1;
+        }
+        
+        if(valuesSoFar == numValues) {
+            sendBufferedValues.schedule();
+        }
+    =}
+    
+    reaction(sendBufferedValues) -> outValue, sendBufferedValues {=
+        
+        if(sendPos < pos) {
+            outValue.set(orderingArray[sendPos]);
+            sendPos += 1;
+            sendBufferedValues.schedule();
+        } else {
+            
+            // reset local state
+            valuesSoFar = 0;
+            pos = 0;
+            sendPos = 0;
+        }
+    =}
+}
+
+reactor ValidationReactor(numValues:int(100000)) {
+    
+    private preamble {=
+        #include "reactor-cpp/logging.hh"
+    =}
+    
+    state sumSoFar:double(0.0);
+    state valuesSoFar:int(0);
+    state prevValue:{=long=}(0);
+    state errorValueLong:{=long=}(-1);
+    state errorValueInt:int(-1);
+    
+    input inValue:{=long=};
+    output outFinished:void;
+    
+    reaction(inValue) -> outFinished {=
+        
+        long value = *inValue.get();
+        valuesSoFar += 1;
+        if(value < prevValue && errorValueLong < 0) {
+            errorValueLong = value;
+            errorValueInt = valuesSoFar - 1;
+        }
+        prevValue = value;
+        sumSoFar += prevValue;
+        
+        if(valuesSoFar == numValues) {
+            if(errorValueLong >= 0) {
+                reactor::log::Error() << "ERROR: Value out of place: " << errorValueLong << " at index " << errorValueInt;
+            } else {
+                reactor::log::Info() << "Elements sum: " << sumSoFar;
+            }
+            
+            // reset local state
+            sumSoFar = 0.0;
+            valuesSoFar = 0;
+            prevValue = 0;
+            errorValueLong = -1;
+            errorValueInt = -1;
+            
+            outFinished.set();
+        }
+    =}
+}
+
+// FIXME We cannot adjust maxValue here because the number of sort actors below depends on it. The parameter should
+// actually indicate the number of sorters, and than we could calculate the max value from it. But this would require
+// changing the original Savina code ... 
+main reactor (numIterations:int(12), dataSize:int(100000), maxValue: long({=1l << 60=}), seed: long({=2048l=})) {
+    
+    source = new IntSourceReactor(numValues=dataSize, maxValue=maxValue, seed=seed);
+    validator = new ValidationReactor(numValues=dataSize);
+    runner = new BenchmarkRunner(numIterations=numIterations);
+    
+    runner.start -> source.inStart;
+    validator.outFinished -> runner.finished;
+    
+    reaction(startup) {=
+        printBenchmarkInfo("RadixSortBenchmark");
+        printArgs("numIterations", numIterations, "dataSize", dataSize, "maxValue", maxValue, "seed", seed);
+        printSystemInfo();
+    =}
+    
+    sorters = new[60] SortActor(numValues=dataSize);
+
+    source.outNextSortReactor, sorters.outValue -> sorters.inValue, validator.inValue;
```

Comment:
This looks like another location where we fail to exploit pipeline parallelism because the connections are all direct feed-through.